src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionBase.cs

```diff
@@ -0,0 +1,212 @@
+
+                var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return true;
+                }
+
+                var service = GetService(document);
+                if (service == null)
+                {
+                    return true;
+                }
+
+                var edits = CollectEdits(document, service, selectedSpans, command, cancellationToken).WaitAndGetResult(cancellationToken);
+
+                ApplyEdits(document, textView, subjectBuffer, service, title, edits);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Applies the requested edits and sets the selection.
+        /// This operation is not cancellable.
+        /// </summary>
+        private void ApplyEdits(Document document, ITextView textView, ITextBuffer subjectBuffer,
+            ICommentSelectionService service, string title, CommentSelectionResult edits)
+        {
+            // Create tracking spans to track the text changes.
+            var currentSnapshot = subjectBuffer.CurrentSnapshot;
+            var trackingSpans = edits.TrackingSpans
+                .Select(textSpan => (originalSpan: textSpan, trackingSpan: CreateTrackingSpan(edits.ResultOperation, currentSnapshot, textSpan.TrackingTextSpan)))
+                .ToImmutableList();
+
+            // Apply the text changes.
+            using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+            {
+                document.Project.Solution.Workspace.ApplyTextChanges(document.Id, edits.TextChanges.Distinct(), CancellationToken.None);
+                transaction.Complete();
+            }
+
+            // Convert the tracking spans into snapshot spans for formatting and selection.
+            var trackingSnapshotSpans = trackingSpans.Select(s => CreateSnapshotSpan(subjectBuffer.CurrentSnapshot, s.trackingSpan, s.originalSpan));
+
+            if (trackingSnapshotSpans.Any())
+            {
+                if (edits.ResultOperation == Operation.Uncomment)
+                {
+                    // Format the document only during uncomment operations.  Use second transaction so it can be undone.
+                    using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                    {
+                        Format(service, subjectBuffer.CurrentSnapshot, trackingSnapshotSpans, CancellationToken.None);
+                        transaction.Complete();
+                    }
+                }
+
+                // Set the selection after the edits have been applied.
+                var spansToSelect = trackingSnapshotSpans.Select(s => new Selection(s));
+                textView.GetMultiSelectionBroker().SetSelectionRange(spansToSelect, spansToSelect.Last());
```

Comment:
why are we only selecting the last span?

Reply:
The second parameter, ```spansToSelect.Last()``` is a single span marking the primary selection.

Reply:
as per something i mentioned earlier, when i give questions like tihs, it's generally an indication that the code should be clearer :)  (i.e. with a comment).  If you'd like, i can be more explicit about this

Reply:
Ahh ok!  Sorry, it's sometimes hard to tell if you're asking for an explanation or asking for a comment :)

Reply:
that's my bad then.  i'll be more explicit :)